EIPS/eip-draft-blake2b-f-precompile.md

```diff
@@ -0,0 +1,76 @@
+status: Draft
+type: Standards Track
+category: Core
+created 2016-10-04
+requires: 2046
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This EIP will enable the Blake2b hash function to run cheaply on the EVM, allowing easier interoperability between Ethereum and Zcash as well as other Equihash-based PoW coins.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+This EIP introduces a new precompiled contract which implements the compression function `F` used in the BLAKE2b cryptographic hashing algorithm, for the purpose of allowing interoperability between the EVM and Zcash, as well as introducing more flexible cryptographic hash primitives to the EVM.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Adds a precompile at address `0x0d` which accepts [ABI encoded](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) arguments corresponding to the function signature
+
+```
+F(bytes32[2] h, bytes32[4] m, uint t , bool f, uint rounds) returns (bytes32[2] h_new);
```

Comment:
Our `F` Go implementation has the following signature:
```
F(h *[8]uint64, blocks []byte, t *[2]uint64, f uint64, rounds uint8)
```
- `h [8]uint64` = `8 * 64 bits = 512 bits = 2 * 32 * 8 bits = 2 * 32 bytes` = `bytes32[2] h`
- `blocks []byte` can fit `bytes32[4] m` but worth to confirm with ZCash if it's enough for all their use cases 
- `t [2]uint64` is not compatible with `uint t`
- `f uint64` is not compatible with `bool f`
- `rounds uint8` is generally compatible with `uint rounds` because we currently allow for `[0,12]` rounds.

What's interesting, the Go implementation from Tjaden [has a similar function signature](https://github.com/tjade273/go-ethereum/blob/39433efa7aff22f426946aab57c44dce88db6fb7/crypto/blake2b/blake2.go#L283) to ours:
```
func BlakeCompress(h [8]uint64, m []uint8, t [2]uint64, f [2]uint64) [8]uint64
```   
I am not sure what is the reason for `t` and `f` to be just `uint/bool` here.

Reply:
After rethinking it:
- ZCash uses both BLAKE2s and BLAKE2b. For 2s, `h` could be represented as `[8]uint32`and for 2b, `h` could be represented as `[8]uint64` on Go side. We are implementing a generic function `F` so it's indeed better to pack the state vector into `bytes32[2]` on Solidity side and use `[8]uint64` on Go side. If needed, two Solidity functions delegating to generic `F` can be exposed later: one with `bytes8[8] h` for 2b and one with `bytes4[8] h` for 2s.
- block size is `128` so `bytes32[4]` fits nicely here on Solidity side. We can even refactor Go `F` implementation to accept `blocks [BlockSize]byte` instead of just `blocks []byte`
- According to the RFC, `f` can be boolean:
    ```
     IF f = TRUE THEN              // last block flag?
       v[14] := v[14] ^ 0xFF..FF   // Invert all bits.
     END IF.
    ```
    I will refactor Go `F` implementation to accept `bool` instead of `uint64` to avoid confusion.
- I still can't agree about `t [2]uint64`. The RFC says `t` is a `2w-bit offset counter`. For 32, we should have `[2]uint32` in Go and for 64 we should have `[2]uint64` in Go. Since `uint` in Solidity defaults to `uint256` we have more than required there. Why not `uint128` in Solidity?

Reply:
Is there actually a benefit from using a `uint218` in Solidity? I believe `uint256` is a single slot on the EVM- so we should always pack as tightly as we can into that base unit.

Reply:
Picking this up in a more public EIP PR

Reply:
> Is there actually a benefit from using a uint128 in Solidity? I believe uint256 is a single slot on the EVM- so we should always pack as tightly as we can into that base unit.

Very good point but I think it matters only when we store data on the chain. Here we just pass parameters so it should be all in-memory.

Reply:
I am going to resolve this conversation since we decided not to ABI-encode the parameters.